src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-server.cpp

```diff
@@ -0,0 +1,191 @@
+    {
+        gDelegateTable[endpoint] = delegate;
+    }
+}
+
+LaundryWasherControlsServer & LaundryWasherControlsServer::Instance()
+{
+    return sInstance;
+}
+
+EmberAfStatus LaundryWasherControlsServer::SetSpinSpeedCurrent(EndpointId endpointId,
+                                                               DataModel::Nullable<uint8_t> newSpinSpeedCurrent)
+{
+    DataModel::Nullable<uint8_t> spinSpeedCurrent;
+    EmberAfStatus res = SpinSpeedCurrent::Get(endpointId, spinSpeedCurrent);
+    if ((res == EMBER_ZCL_STATUS_SUCCESS) && (spinSpeedCurrent != newSpinSpeedCurrent))
+    {
+        res = SpinSpeedCurrent::Set(endpointId, newSpinSpeedCurrent);
+    }
+
+    return res;
+}
+
+EmberAfStatus LaundryWasherControlsServer::GetSpinSpeedCurrent(EndpointId endpointId,
+                                                               DataModel::Nullable<uint8_t> & spinSpeedCurrent)
+{
+    return SpinSpeedCurrent::Get(endpointId, spinSpeedCurrent);
+}
+
+EmberAfStatus LaundryWasherControlsServer::SetNumberOfRinses(EndpointId endpointId, NumberOfRinsesEnum newNumberOfRinses)
+{
+    NumberOfRinsesEnum numberOfRinses;
+    EmberAfStatus res = NumberOfRinses::Get(endpointId, &numberOfRinses);
+
+    if ((res == EMBER_ZCL_STATUS_SUCCESS) && (numberOfRinses != newNumberOfRinses))
+    {
+        res = NumberOfRinses::Set(endpointId, newNumberOfRinses);
+    }
+
+    return res;
+}
+
+EmberAfStatus LaundryWasherControlsServer::GetNumberOfRinses(EndpointId endpointId, NumberOfRinsesEnum & numberOfRinses)
+{
+    return NumberOfRinses::Get(endpointId, &numberOfRinses);
+}
+
+/**********************************************************
+ * LaundryWasherControlsServer private methods
+ *********************************************************/
+CHIP_ERROR LaundryWasherControlsServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    if (aPath.mClusterId != LaundryWasherControls::Id)
+    {
+        // We shouldn't have been called at all.
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    switch (aPath.mAttributeId)
+    {
+    case Attributes::SpinSpeeds::Id:
+        return ReadSpinSpeeds(aPath, aEncoder);
+    case Attributes::SupportedRinses::Id:
+        return ReadSupportedRinses(aPath, aEncoder);
+    default:
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR LaundryWasherControlsServer::ReadSpinSpeeds(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    aEncoder.EncodeEmptyList();
+    Delegate * delegate = GetDelegate(aPath.mEndpointId);
+    VerifyOrReturnError(delegate != nullptr, CHIP_ERROR_INCORRECT_STATE, ChipLogError(Zcl, "Delegate is nullptr"));
+
+    return aEncoder.EncodeList([&](const auto & encoder) -> CHIP_ERROR {
+        for (uint8_t i = 0; true; i++)
+        {
+            CharSpan spinSpeed;
+            auto err = delegate->GetSpinSpeedAtIndex(i, spinSpeed);
+            if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
```

Comment:
OK, so this is relying on CHIP_ERROR_PROVIDER_LIST_EXHAUSTED!  Please document that in the relevant API documentation.

Reply:
I will add it